### PR TITLE
[Reviewer: Graeme] Coverage fixes

### PR DESCRIPTION
**Pull request by eleanor-merry**
This PR tidies up the code a bit, fixes up the behaviour if you use the wildcard contact header, and improves the general UT coverage.

### CODE REVIEW

File: src/registrarsproutlet.cpp [hunk 1]
```diff
@@ -669,6 +663,16 @@ void RegistrarSproutletTsx::get_bindings_from_req(
 
   while (contact != NULL)
   {
+    if (contact->star)
```
[the remark this answers is not in the document]
Comment (graemerobertson) [reply src/registrarsproutlet.cpp]: So the contact is a wildcard or something? Not sure why that makes this code the right thing to do?! Maybe a comment?

File: src/ut/hssconnection_test.cpp [hunk 2]
```diff
@@ -446,6 +447,16 @@ TEST_F(HssConnectionTest, SimpleNotRegisteredGet)
...
 }
 
+TEST_F(HssConnectionTest, GetServerFailure)
+{
+  HSSConnection::irs_query irs_query;
+  HSSConnection::irs_info irs_info;
+
+  HTTPCode rc = _hss.get_registration_data("serverfailureonget", irs_info, 0);
+
+  EXPECT_EQ(rc, 404);
```
Comment (graemerobertson) [inline src/ut/hssconnection_test.cpp]: 404 seems odds for a "server failure"?

File: src/ut/registrar_test.cpp [hunk 3]
```diff
@@ -618,6 +618,41 @@ TEST_F(RegistrarTest, DeRegisterSubscriber)
   EXPECT_EQ(1,((SNMP::FakeSuccessFailCountTable*)SNMP::FAKE_REGISTRATION_STATS_TABLES.de_reg_tbl)->_successes);
 }
 
+// Test deregistering a subscriber with the wildcard contact header.
+TEST_F(RegistrarTest, DeRegisterSubscriberWithWildcard)
```
Comment (graemerobertson) [inline src/ut/registrar_test.cpp]: DeRegister?! Gross!! Ugh it's everywhere.

Comment (graemerobertson) [reply src/ut/registrar_test.cpp]: I'm excited about this test though, maybe it'll help me understand the code you wrote earlier in the PR

Comment (graemerobertson) [reply src/ut/registrar_test.cpp]: Nope

File: src/ut/s4_test.cpp [hunk 4]
```diff
@@ -1016,6 +1044,45 @@ TEST_F(BasicS4Test, ClearUpEmptyAoR)
...
+// subscriptions are removed.
+TEST_F(BasicS4Test, ClearUpSubscriptionsInAoR)
+{
+  AoR* get_aor = AoRTestUtils::create_simple_aor("aor_id");
+  get_aor->get_binding(AoRTestUtils::BINDING_ID)->_emergency_registration = true;
+  Binding* binding = AoRTestUtils::build_binding("aor_id", time(NULL));
+  get_aor->_bindings.insert(std::make_pair(AoRTestUtils::BINDING_ID + "2", binding));
+  std::string get_str = _serializer_deserializer.serialize_aor(get_aor);
+
+  AoR* expect_aor = AoRTestUtils::create_simple_aor("aor_id", false);
```
Comment (graemerobertson) [inline src/ut/s4_test.cpp]: It took me a while to figure out what was going on here. I now assume that passing `false` to `create_simple_aor` here creates an AoR without any subscriptions?! If so, worth explicitly commenting that probably

File: src/ut/sprout_xml_utils_test.cpp [hunk 5]
```diff
@@ -15,17 +15,97 @@
...
+  SproutXMLUtilsTest()
+  {
+  }
+
+  virtual ~SproutXMLUtilsTest()
+  {
+  }
 
-/*std::string COM = ""
+  void check_expectations(std::string xml,
```
Comment (graemerobertson) [inline src/ut/sprout_xml_utils_test.cpp]: Pass some of these parameters by reference? Pft, not sure why I'm bothering commenting on this in UT code.

File: src/ut/sprout_xml_utils_test.cpp [hunk 6]
```diff
@@ -15,17 +15,97 @@
...
+    {
+      EXPECT_TRUE(exp_associated_uris == associated_uris);
+      std::sort(aliases.begin(), aliases.end());
+      std::sort(exp_aliases.begin(), exp_aliases.end());
+      EXPECT_EQ(exp_aliases, aliases);
+    }
+
+    free(cstr_xml);
+  }
+
```
Comment (graemerobertson) [inline src/ut/sprout_xml_utils_test.cpp]: I'd get rid of this line

File: src/registrarsproutlet.cpp (hunk 1 above)
[the remark this answers is not in the document]
Comment (eleanor-merry) [reply src/registrarsproutlet.cpp]: It's a contact header that looks like "Contact: *", and it means that we should remove all existing bindings. I've added a comment.

File: src/registrarsproutlet.cpp [hunk 7]
```diff
@@ -505,19 +505,13 @@ void RegistrarSproutletTsx::process_register_request(pjsip_msg *req)
...
   {
-    int max_expiry = AoRUtils::get_max_expiry(all_bindings, now);
-
-    std::string as_reg_id = public_id;
+    pjsip_msg* clone_rsp = clone_msg(rsp);
 
-    if (irs_info._associated_uris.is_impu_barred(public_id))
-    {
-      as_reg_id = default_impu;
-    }
```
[the remark this answers is not in the document]
Comment (eleanor-merry) [reply src/registrarsproutlet.cpp]: It's an impossible case. We shouldn't send third party registers with a barred public identity, but we can never have reached this point in the code if it was.

File: src/registrarsproutlet.cpp [hunk 8]
```diff
@@ -980,7 +984,7 @@ void RegistrarSproutletTsx::add_contact_headers(pjsip_msg* rsp,
 
       // Add a GRUU if the UE supports GRUUs and the contact header contains
       // a +sip.instance parameter.
-      if (PJUtils::msg_supports_extension(req, "gruu")) // EM-TODO should be a constant
+      if (PJUtils::msg_supports_extension(req, "gruu"))
```
[the remark this answers is not in the document]
Comment (eleanor-merry) [reply src/registrarsproutlet.cpp]: Nailed it (but it's existing code, and I think we've run out of time to make unrelated fixes).

File: src/ut/hssconnection_test.cpp (hunk 2 above)
Comment (eleanor-merry) [reply src/ut/hssconnection_test.cpp]: Good point - it's just a really badly named test. I've renamed it.

**Pull request merged**